src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,316 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Experiments;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _annotation = new SyntaxAnnotation();
+
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract ImmutableHashSet<string> GetNamespacesInScope(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        private readonly ITypeImportCompletionService _typeImportCompletionService;
+        private readonly IExperimentationService _experimentationService;
+
+        public AbstractTypeImportCompletionProvider(Workspace workspace)
+            : base()
+        {
+            _typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            _experimentationService = workspace.Services.GetService<IExperimentationService>();
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+#if DEBUG
+            DebugObject.DebugClear();
+            var tick = System.Environment.TickCount;
+#endif
+            var document = completionContext.Document;
+            var position = completionContext.Position;
+            var cancellationToken = completionContext.CancellationToken;
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && _experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var items = await GetCompletionItemsAsync(document, syntaxContext, position, cancellationToken).ConfigureAwait(false);
+            completionContext.AddItems(items);
```

Comment:
consider just passing the completion context to GetCompletionItemsAsync (and call it AddCompletionItemsAsync).  So no need to allocate the large collection

Reply:
Fixed